homeassistant/components/remote/harmony.py

```diff
@@ -176,60 +187,51 @@ def is_on(self):
 
-    def update(self):
-        """Return current activity."""
+    def new_activity(self, activity_id):
+        """Callback for updating the current activity."""
         import pyharmony
-        name = self._name
-        _LOGGER.debug("Polling %s for current activity", name)
-        state = pyharmony.ha_get_current_activity(
-            self._token, self._config, self.host, self._port)
-        _LOGGER.debug("%s current activity reported as: %s", name, state)
-        self._current_activity = state
-        self._state = bool(state != 'PowerOff')
+        activity_name = pyharmony.activity_name(self._config, activity_id)
+        _LOGGER.debug("%s activity reported as: %s", self._name, activity_name)
+        self._current_activity = activity_name
+        self._state = bool(self._current_activity != 'PowerOff')
+        self.schedule_update_ha_state()
 
     def turn_on(self, **kwargs):
         """Start an activity from the Harmony device."""
         import pyharmony
-        if kwargs[ATTR_ACTIVITY]:
-            activity = kwargs[ATTR_ACTIVITY]
-        else:
-            activity = self._default_activity
+        activity = kwargs.get(ATTR_ACTIVITY, self._default_activity)
 
         if activity:
-            pyharmony.ha_start_activity(
-                self._token, self.host, self._port, self._config, activity)
+            activity_id = pyharmony.activity_id(self._config, activity)
+            self._client.start_activity(activity_id)
             self._state = True
         else:
             _LOGGER.error("No activity specified with turn_on service")
 
     def turn_off(self, **kwargs):
         """Start the PowerOff activity."""
-        import pyharmony
-        pyharmony.ha_power_off(self._token, self.host, self._port)
+        self._client.power_off()
 
-    def send_command(self, command, **kwargs):
-        """Send a set of commands to one device."""
-        import pyharmony
-        device = kwargs.pop(ATTR_DEVICE, None)
+    def send_command(self, commands, **kwargs):
+        """Send a list of commands to one device."""
+        device = kwargs.get(ATTR_DEVICE)
         if device is None:
             _LOGGER.error("Missing required argument: device")
             return
-        params = {}
-        num_repeats = kwargs.pop(ATTR_NUM_REPEATS, None)
-        if num_repeats is not None:
-            params['repeat_num'] = num_repeats
-        delay_secs = kwargs.pop(ATTR_DELAY_SECS, None)
-        if delay_secs is not None:
-            params['delay_secs'] = delay_secs
-        pyharmony.ha_send_commands(
-            self._token, self.host, self._port, device, command, **params)
+
+        num_repeats = kwargs.get(ATTR_NUM_REPEATS)
+        delay_secs = kwargs.get(ATTR_DELAY_SECS, self._delay_secs)
+
+        for _ in range(num_repeats):
+            for command in commands:
+                self._client.send_command(device, command)
+                time.sleep(delay_secs)
```

Comment:
I don't think you should sleep occupying a thread. Probably schedule a call in the future instead.

Reply:
That does sound better. I made it like this for two reasons: a) to ensure that the ordering is kept if a script has multiple `send_command` calls and b) to allow short delays like 0.4 seconds. Can I meet these goals when scheduling a call?

(This is not a regression as the existing call through pyharmony also ends up doing a sleep.)

Reply:
>This is not a regression as the existing call through pyharmony also ends up doing a sleep.

Sleeping in a separate thread that is not part of the home assistant asyncio thread pool, will not risk interfering with the performance of the thread pool. So I don't think it's the same cases.

Reply:
Already today it sleeps in a home assistant thread ("SyncWorker_N").